src/other/lam-7.1.4/man/man1/lamexec.1
.TH LAMEXEC 1 "July, 2007" "LAM 7.1.4" "LAM COMMANDS"
.SH NAME
lamexec \- Run non-MPI programs on LAM nodes.
.SH SYNOPSIS
.hy 0
.HP
.na
lamexec 
[-fhvD] 
[-c # | -np #] 
[-nw | -w] 
[-pty]
[-s node] 
[-x VAR1[=VALUE1][,VAR2[=VALUE2],...]]
[where] 
program [-- args]
.br
.RE
.ad b
.hy 1
.SH OPTIONS
.TP 10
.B -c #
Synonym for 
.I \-np 
(see below).
.TP
.B -D
Use the executable program location as the current working directory
for created processes.  The current working directory of the created
processes will be set before the user's program is invoked.
.TP
.B -f
Do not configure standard I/O file descriptors - use defaults.
.TP
.B -h
Print useful information on this command.
.TP
.B -np #
(see below).
Run this many copies of the program on the given nodes.  This option
indicates that the specified file is an executable program and not an
application schema.  If no nodes are specified, all LAM nodes are
considered for scheduling; LAM will schedule the programs in a
round-robin fashion, "wrapping around" (and scheduling multiple copies
on a single node) if necessary.
.TP
.B -nw
Do not wait for all processes to complete before exiting
.IR lamexec .
This option is mutually exclusive with 
.IR \-w .
.TP
.B -pty
Enable pseudo-tty support.  Among other things, this enabled
line-buffered output (which is probably what you want).  The only
reason that this feature is not enabled by default is because it is so
new and has not been extensively tested yet.
.TP
.B -s node
Load the program from this node.  This option is not valid on the
command line if an application schema is specified.
.TP
.B -v
Be verbose; report on important steps as they are done.
.TP
.B -w
Wait for all applications to exit before 
.IR lamexec
exits.
.TP 
.B -x
Export the specified environment variables to the remote nodes before
executing the program.  Existing environment variables can be
specified (see the Examples section, below), or new variable names
specified with corresponding values.  The parser for the 
.I \-x 
option is not very sophisticated; it does not even understand quoted
values.  Users are advised to set variables in the environment, and
then use
.I \-x 
to export (not define) them.
.TP
.B where
A set of node and/or CPU identifiers indicating where to start 
.TP
.B -- args
Pass these runtime arguments to every new process.  This must always
be the last argument to
.IR lamexec .
This option is not valid on the command line if an application schema
is specified.
.SH DESCRIPTION
.I lamexec
is essentially a clone of the 
.IR mpirun (1), 
but is intended for non-MPI programs.
.PP
One invocation of
.I lamexec
starts a non-MPI application running under LAM.  To start the same
program on all LAM nodes, the application can be specified on the
.I lamexec
command line.  To start multiple applications on the LAM nodes, an
application schema is required in a separate file.  See appschema(5)
for a description of the application schema syntax, but it essentially
contains multiple
.I lamexec
command lines, less the command name itself.  The ability to specify
different options for different instantiations of a program is another
reason to use an application schema.
.SS Location Nomenclature
The location nomenclature that is used for the where clause mention
in the SYNOPSIS section, above, is identical to mpirun(1)'s
nomenclature.  See the mpirun(1) man page for a lengthy discussion of
the location nomenclature.
.PP
Note that the by-CPU syntax, while valid for
.IR lamexec ,
is not quite as meaningful because process rank ordering in
MPI_COMM_WORLD is irrelevant.  As such, the by-node nomenclature is
typically the preferred syntax for 
.IR lamexec .
.SS Application Schema or Executable Program?
To distinguish the two different forms,
.I lamexec
looks on the command line for nodes or the \fI-c\fR option.  If
neither is specified, then the file named on the command line is
assumed to be an application schema.  If either one or both are
specified, then the file is assumed to be an executable program.  If
nodes and \fI-c\fR both are specified, then copies of the program
are started on the specified nodes according to an internal LAM
scheduling policy.  Specifying just one node effectively forces LAM to
run all copies of the program in one place.  If \fI-c\fR is given, but
not nodes, then all LAM nodes are used.  If nodes is given, but
not \fI-c\fR, then one copy of the program is run on each node.
.SS Program Transfer
By default, LAM searches for executable programs on the target node
where a particular instantiation will run.  If the file system is not
shared, the target nodes are homogeneous, and the program is
frequently recompiled, it can be convenient to have LAM transfer the
program from a source node (usually the local node) to each target
node.  The \fI-s\fR option specifies this behavior and identifies the
single source node.
.SS Locating Files
LAM looks for an executable program by searching the directories in
the user's PATH environment variable as defined on the source node(s).
This behavior is consistent with logging into the source node and
executing the program from the shell.  On remote nodes, the "." path
is the home directory.
.PP
LAM looks for an application schema in three directories: the local
directory, the value of the LAMAPPLDIR environment variable, and
laminstalldir/boot, where "laminstalldir" is the directory where
LAM/MPI was installed.
.SS Standard I/O
LAM directs UNIX standard input to /dev/null on all remote nodes.  On
the local node that invoked
.IR lamexec ,
standard input is inherited from
.IR lamexec .
The default is what used to be the -w option to prevent conflicting
access to the terminal.
.PP
LAM directs UNIX standard output and error to the LAM daemon on all
remote nodes.  LAM ships all captured output/error to the node that
invoked
.I lamexec
and prints it on the standard output/error of
.IR lamexec .
Local processes inherit the standard output/error of
.I lamexec
and transfer to it directly.
.PP
Thus it is possible to redirect standard I/O for LAM applications by
using the typical shell redirection procedure on
.IR lamexec .
.sp
.RS
% lamexec N my_app  my_input  my_output
.RE
.PP
The
.I \-f
option avoids all the setup required to support standard I/O described
above.  Remote processes are completely directed to /dev/null and
local processes inherit file descriptors from lamboot(1).
.SS Pseudo-tty support
The 
.I \-pty
option enabled pseudo-tty support for process output.  This allows,
among other things, for line buffered output from remote nodes (which
is probably what you want).  
.PP
This option is not currently the default for 
.IR lamexec 
because it has not been thoroughly tested on a variety of different
Unixes. Users are encouraged to use
.I \-pty 
and report any problems back to the LAM Team.
.SS Current Working Directory
The current working directory for new processes created on the local
node is inherited from
.IR lamexec .
The current working directory for new processes created on remote
nodes is the remote user's home directory.  This default behavior is
overridden by the
.I \-D
option.  
.P
The
.I \-D
option will change the current working directory of new processes to
the directory where the executable resides before the new user's
program is invoked.  
.PP
An alternative to the 
.I \-D 
option is the
.I \-wd
option.  
.I \-wd
allows the user to specify an arbitrary current working directory
(vs. the location of the executable).  Note that the 
.I \-wd
option can be used in application schema files (see appschema(5)) as
well.
.SS Process Environment
Processes in the application inherit their environment from the LAM
daemon upon the node on which they are running.  The environment of a
LAM daemon is fixed upon booting of the LAM with lamboot(1) and is
inherited from the user's shell.  On the origin node this will be the
shell from which lamboot(1) was invoked and on remote nodes this will
be the shell started by rsh(1).  When running dynamically linked
applications which require the LD_LIBRARY_PATH environment variable to
be set, care must be taken to ensure that it is correctly set when
booting the LAM.
.SS Exported Environment Variables
The 
.I \-x
option to 
.IR lamexec
can be used to export specific environment variables to the new
processes.  While the syntax of the 
.I \-x 
option allows the definition of new variables, note that the parser
for this option is currently not very sophisticated - it does not even
understand quoted values.  Users are advised to set variables in the
environment and use 
.I \-x 
to export them; not to define them.
.SH EXAMPLES
.TP 4
lamexec N prog1
Load and execute prog1 on all nodes.
Search for the executable file on each node.
.TP
lamexec -c 8 prog1
Run 8 copies of prog1 wherever LAM wants to run them.
.TP
lamexec n8-10 -v -nw -s n3 prog1 -- -q
Load and execute prog1 on nodes 8, 9, and 10.  Search for prog1 on
node 3 and transfer it to the three target nodes.  Report as each
process is created.  Give "-q" as a command line to each new process.
Do not wait for the processes to complete before exiting
.IR lamexec .
.TP
lamexec -v myapp
Parse the application schema, myapp, and start all processes specified
in it.  Report as each process is created.
.TP
lamexec N N -pty -wd /workstuff/output -x DISPLAY run_app.csh
Run the application "run_app.csh" (assumedly a C shell script) twice
on each node in the system (ideal for 2-way SMPs).  Also enable
pseudo-tty support, change directory to /workstuff/output, and export
the DISPLAY variable to the new processes (perhaps the shell script
will invoke an X application such as xv to display output).
.TP
lamexec -np 5 -D `pwd`/my_application
A common usage of 
.IR lamexec
in environments where a filesystem is shared between all nodes in the
multicomputer, using the shell-escaped "pwd" command specifies the
full name of the executable to run.  This prevents the need for
putting the directory in the path; the remote notes will have an
absolute filename to execute (and change directory to it upon
invocation).
.SH DIAGNOSTICS
.TP 4
lamexec: Exec format error
A non-ASCII character was detected in the application schema.
This is usually a command line usage error where
.I lamexec
is expecting an application schema and an executable file was given.
.TP
lamexec: syntax error in application schema, line XXX
The application schema cannot be parsed because of a usage or syntax error
on the given line in the file.
.TP
filename: No such file or directory
This error can occur in two cases.  Either the named file cannot be
located or it has been found but the user does not have sufficient
permissions to execute the program or read the application schema.
.SH RETURN VALUE
.I lamexec
returns 0 if all processes started by
.I lamexec
exit normally.  A non-zero value is returned if an internal error occurred 
in lamexec, or one or more processes exited abnormally.  If an internal 
error occurred in lamexec, the corresponding error code is returned.  In 
the event that one or more processes exit with non-zero exit code, the 
return value of the process that
.I lamexec
first notices died abnormally will be returned.  Note that, in general, 
this will be the first process that died but is not guaranteed to be so.
.PP
However, note that if the
.I \-nw
switch is used, the return value from lamexec does not indicate the exit 
status of the processes started by it.
.SH SEE ALSO
mpimsg(1), mpirun(1), mpitask(1), loadgo(1)


src/other/lam-7.1.4/otb/lamexec/lamexec.c
/*
 * Copyright (c) 2001-2006 The Trustees of Indiana University.  
 *                         All rights reserved.
 * Copyright (c) 1998-2001 University of Notre Dame. 
 *                         All rights reserved.
 * Copyright (c) 1994-1998 The Ohio State University.  
 *                         All rights reserved.
 * 
 * This file is part of the LAM/MPI software package.  For license
 * information, see the LICENSE file in the top level directory of the
 * LAM/MPI source distribution.
 * 
 * $HEADER$
 *
 * $Id: lamexec.c,v 1.26 2003/11/15 05:40:53 pkambadu Exp $
 *
 *	Function:	- run commands on a LAM
 */

#include <lam_config.h>

#include <sys/types.h>
#include <sys/signal.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/param.h>
#include <netdb.h>
#include <unistd.h>

#include <all_list.h>
#include <all_opt.h>
#include <app_mgmt.h>
#include <app_schema.h>
#include <args.h>
#include <events.h>
#include <freq.h>
#include <kio.h>
#include <laminternal.h>
#include <ndi.h>
#include <net.h>
#include <portable.h>
#include <priority.h>
#include <preq.h>
#include <sfh.h>
#include <terror.h>
#include <typical.h>
#include <etc_misc.h>

/*
 * private functions
 */
static int pwait(int4 nwait, int *childstat);
static int set_mode(void);
static int set_stdio(void);
static void reset_stdio(void);
static void first_signal_handler(int sig); 
static void signal_handler(int sig);
static LIST *build_job(int argc, char **argv, char **env);
static char *locate_aschema(char *filename);
static void node_ranges(char *buf, int4 initnode, int4 initindex);
static void error_cleanup(void);

/*
 * external variables
 */
extern struct kio_t	_kio;			/* kernel I/O block */
extern struct fclient	_ufd[FUMAX];		/* user file desc. */

/*
 * private variables
 */
static struct _gps	*job;			/* job GPS array */
static int		job_n;			/* size of job */
static int		fl_fdesc;		/* don't open stdio */
static int		fl_nprocs;		/* # procs specified */
static int		fl_verbose;		/* verbose mode */
static int		fl_wait = 1;		/* wait for completion */
static int4		nprocs;			/* # procs */
static int4		rtf;			/* runtime flags */
static char		*wrkdir;		/* working dir */
static char		fmtbuf[512];		/* formatting buffer */
static char		smallbuf[128];		/* small extra buffer */
static OPT		*ad;			/* argument descriptor */

static char		*syntaxerr =
	"lamexec: application schema syntax error, line %d\n";

static char		*usage =
	"lamexec [options] [<app>] [[<nodes>] <prog> [<args>]]\n";

int
main(int argc, char **argv)
{
    LIST	*app;			/* application desc. */
    int		status;			/* return status */
    char	**env = 0;
/*
 * Parse the command line.
 */
    ad = ao_init();
    if (ad == 0) {
	perror("lamexec (ao_init)");
	exit(errno);
    }
    ao_setopt1(ad, "fhvwD", 0, 0, 0);
    ao_setopt(ad, "s", 0, 1, 0);
    ao_setopt(ad, "c", 0, 1, AOINT);
    ao_setopt(ad, "np", "c", 1, AOINT);
    ao_setopt(ad, "nw", "w", 0, 0);
    ao_setopt(ad, "x", 0, 1, 0);
    ao_setopt(ad, "wd", "D", 1, 0);
    ao_setopt(ad, "pty", 0, 0, 0);
    ao_setopt(ad, "npty", "pty", 0, 0);

    if (asc_compat(&argc, &argv, ad)) {
	perror("lamexec");
	exit(errno);
    }

    if (ao_parse(ad, &argc, argv))  {
	fprintf(stderr, usage);
	exit(errno);
    }
/*
 * Check for help request.
 */
    if (ao_taken(ad, "h")) {
      show_help("lamexec", "usage", NULL);
      exit(0);
    }
/*
 * Set the operating mode.
 */
    if (set_mode()) {
	terror("lamexec (set_mode)");
	exit(errno);
    }
/*
 * Attach to kernel.
 */
    if (kinit(PRCMD)) {
      char hostname[MAXHOSTNAMELEN];
      gethostname(hostname, MAXHOSTNAMELEN);

      show_help(NULL, "no-lamd", "lamexec", hostname, NULL);
      exit(errno);
    }
/*
 * Set job identifier to be inherited by the programs.
 */
    _kio.ki_jobid.jid_node = getnodeid();
    _kio.ki_jobid.jid_pid = lam_getpid();
/*
 * Get the job environment.
 */
    if (asc_environment(0,
	    ao_taken(ad, "x") ? ao_param(ad, "x", 0, 0) : NULL, &env))
	lamfail("lamexec");
/*
 * Build the job.
 */
    if ((app = build_job(argc, argv, env)) == 0) {
	kexit(errno);
    }

    job_n = al_count(app);
    if ((job = malloc(job_n * sizeof(struct _gps))) == 0)
	lamfail("lamexec");
  
/* We need to have different signal handlers till we can cleanup all
 * the socket structures we need. This is a vulnerable period to recieve
 * signals. Some sockets may remain open. This signal handler will generate
 * a warning on the first occurance of the signal and on the second occurance
 * it will call the generic "signal_handler()"
 */
 if (fl_wait) {
   _lam_signal(SIGINT, first_signal_handler);
	 _lam_signal(SIGTERM, first_signal_handler);
 }

/*
 * Set job stdio.
 */
    if (set_stdio())
	lamfail("lamexec (set_stdio)");
/*
 * Run the job.
 */
    if (asc_run(app, 0, rtf, fl_verbose, TRUE, job)) {
	reset_stdio();
	kexit(errno);
    }
/*
 * Set signal handler to terminate the job.
 * Now we change the signal handler to the original one since the 
 * vulnerable period is over
 */
    if (fl_wait) {
	_lam_signal(SIGINT, signal_handler);
	_lam_signal(SIGTERM, signal_handler);
    }

    reset_stdio();
/*
 * If needed, wait for the job to terminate.
 */
    status = 0;

    if (fl_wait) {
	if (pwait(job_n, &status)) {
	    app_doom(job_n, job, -15);
	    error_cleanup();
	}
    }

    kexit(status);
    return(0);
}

/*
 *	build_job
 *
 *	Function:	- build job from command line specification
 *	Accepts:	- command line argument count
 *			- command line argument vector
 *	Returns:	- application descriptor or 0
 */
static LIST *
build_job(int argc, char **argv, char **env)
{
	LIST		*app;			/* application */
	LIST		*app_sched;		/* scheduled application */
	int4		nodeid;			/* target node */
	int4		index;			/* for getnode() */
	int4		nflags;			/* for getnode() */
	int		lineno;			/* line number */
	int		argtailc;		/* argument tail count */
	char		**argtailv;		/* argument tail vector */
	char		*aschema;		/* application schema */
	char		*qarg;			/* quoted argument */

	if (nid_parse(&argc, argv) || (errno = (argc > 1) ? 0 : EUSAGE)) {
		fprintf(stderr, usage);
		return(0);
	}
/*
 * Application schema or command line?
 */
	nid_get(&index, &nodeid, &nflags);

	if ((index < 0) && !fl_nprocs) {
/*
 * Parse the app. schema.
 */
		if ((argc != 2) || ao_taken(ad, "s")) {
			fprintf(stderr, usage);
			kexit(EUSAGE);
		}

		aschema = locate_aschema(argv[1]);
		if (aschema == 0) {
			fprintf(stderr, "lamexec (locate_aschema): %s: ",
				argv[1]);
			terror("");
			return(0);
		}

		app = asc_parse(aschema, &lineno, env);
	}
	else {
/*
 * Parse the "command line" application specification.
 * Reconstruct its relevant parts from the parsed information.
 * It was parsed to distinguish between the appl/cmdline cases.
 *
 * Format: [<nodes>] program [-s <srcnode>] [-c #] [-- args]
 *
 */
		fmtbuf[0] = 0;

		if (fl_nprocs) {
			sprintf(smallbuf, "-c %d ", nprocs);
			strcat(fmtbuf, smallbuf);
		}

		if (ao_taken(ad, "s")) {
			strcat(fmtbuf, "-s ");
			strcat(fmtbuf, ao_param(ad, "s", 0, 0));
			strcat(fmtbuf, " ");
		}

		if (index == 0) {
			if (nodeid == LOCAL) nodeid = getnodeid();
			node_ranges(fmtbuf, nodeid, index);
		}

		for (--argc, ++argv; argc > 0; --argc, ++argv) {
			qarg = sfh_argv_quote(*argv, "\\'\"");
			if (qarg == 0) {
				terror("lamexec (sfh_argv_quote)");
				return(0);
			}
			strcat(fmtbuf, qarg);
			strcat(fmtbuf, " ");
			free(qarg);
		}

		ao_tail(ad, &argtailc, &argtailv);
		if (argtailc > 0) strcat(fmtbuf, "-- ");

		for ( ; argtailc > 0; --argtailc, ++argtailv) {
			qarg = sfh_argv_quote(*argtailv, "\\'\"");
			if (qarg == 0) {
				terror("lamexec (sfh_argv_quote)");
				return(0);
			}
			strcat(fmtbuf, qarg);
			strcat(fmtbuf, " ");
			free(qarg);
		}
/*
 * Parse the command line.
 */
		app = asc_bufparse(fmtbuf, strlen(fmtbuf), &lineno, env);
		lineno = 0;
	}

	if (app == 0) {
		if (lineno > 0) {
			fprintf(stderr, syntaxerr, lineno);
			errno = EUSAGE;
		} else {
			fprintf(stderr, "lamexec: cannot parse: ");
			terror("");
		}

		return(0);
	}
/*
 * Set working dir.
 */
	if (wrkdir) {
	    if (asc_workdir(app, wrkdir)) {
		terror("lamexec (asc_workdir)");
		return(0);
	    }
	}
/*
 * Schedule the application.
 */
	app_sched = asc_schedule(app);
	asc_free(app);

	if (app_sched == 0) {
		terror("lamexec (asc_schedule)");
		return(0);
	} else {
		return(app_sched);
	}
}

/*
 *	pwait
 *
 *	Function:	- waits for processes to complete
 *	Accepts:	- # of processes
 *			- return status of first child in error, else 0
 *	Returns:	- 0 or LAMERROR
 */
static int
pwait(int4 nwait, int *childstat)
{
	int4		nodeid;			/* child's node ID */
	int		pid;			/* child's process ID */
	int		status;			/* return status */

	*childstat = 0;

	for ( ; nwait > 0; --nwait) {

		if (rpwait(&nodeid, &pid, &status)) {
			*childstat = errno;
			terror("lamexec (rpwait)");
			return(LAMERROR);
		}

		status &= ~(_LAM_WIFNODETACH);
		if (status) {
		    if (LAM_WIFSIGNALED(status)) {
			printf("%d (n%d) exited due to signal %d\n",
				pid, nodeid, LAM_WTERMSIG(status));
			*childstat = -1;
		    } else {
			status = LAM_WEXITSTATUS(status);
			printf("%d (n%d) exited with status %d\n",
				pid, nodeid, status);
			*childstat = status;
		    }
		    return(LAMERROR);
		}
	}

	return(0);
}

/*
 *	set_mode
 *
 *	Function:	- set the operating mode
 */
static int
set_mode(void)
{
/*
 * flags
 */
    fl_fdesc = ao_taken(ad, "f");
    fl_verbose = ao_taken(ad, "v");
    if (ao_taken(ad, "nw"))
	fl_wait = 0;
/*
 * followed options
 */
    nprocs = -1;
    if (ao_taken(ad, "c")) {
	ao_intparam(ad, "c", 0, 0, &nprocs);
	fl_nprocs = 1;
    } else if (ao_taken(ad, "np")) {
	ao_intparam(ad, "np", 0, 0, &nprocs);
	fl_nprocs = 1;
    }
    
    if (ao_taken(ad, "wd")) {
	wrkdir = ao_param(ad, "wd", 0, 0);
    }
/*
 * runtime flags
 */
    if (fl_wait)
	rtf |= RTF_WAIT;
    if (ao_taken(ad, "D"))
	rtf |= RTF_APPWD;
    if (ao_taken(ad, "O"))
	rtf |= RTF_HOMOG;
    if (isatty(1))
	rtf |= RTF_TTYOUT;
    if (!fl_fdesc)
	rtf |= RTF_IO;
    if (!ao_taken(ad, "npty"))
	rtf |= RTF_PTYS;

    return(0);
}

/*
 *	set_stdio
 *
 *	Function:	- set up application stdio
 *	Returns:	- 0 or LAMERROR
 */
static int
set_stdio(void)
{
	char		server[LAM_PATH_MAX];	/* fd server socket name */

	if (fl_fdesc) return(0);

/*
 * Pass stdin, stdout and stderr to filed.
 */
	if (lam_mktmpid((int) lam_getpid(), server, sizeof(server))) {
		return(LAMERROR);
	}

	if (lam_lfopenfd(server)) {
		return(LAMERROR);
	}
/*
 * Set LAM file descriptors to the passed file descriptors.  The call to
 * lam_lfopenfd() takes care of the case when stdin is a tty.
 */
	_kio.ki_stdin = _ufd[0].fu_tfd;
	_kio.ki_stdout = _ufd[1].fu_tfd;
	_kio.ki_stderr = _ufd[2].fu_tfd;
	return(0);
}

/*
 *	reset_stdio
 *
 *	Function:	- reset stdio so rfatexit will clean it up
 *	Returns:	- 0 or LAMERROR
 */
static void
reset_stdio(void)
{
	_kio.ki_stdin = 0;
	_kio.ki_stdout = 1;
	_kio.ki_stderr = 2;
}

/*
 *	locate_aschema
 *
 *	Function:	- locate an application schema
 *	Accepts:	- filename
 *	Returns:	- full pathname or NULL
 */
static char *
locate_aschema(char *filename)
{
	int		pathc = 0;		/* paths argc */
	char		**pathv = 0;		/* paths argv */
	char		*appdir;		/* application dir */
	char		*fullpath;		/* full pathname */

	if ((appdir = getenv("LAMAPPLDIR"))) {
		sfh_argv_add(&pathc, &pathv, appdir);
	}
	sfh_argv_add(&pathc, &pathv, "");
	sfh_argv_add(&pathc, &pathv, "$LAMHOME/etc");
	sfh_argv_add(&pathc, &pathv, "$TROLLIUSHOME/etc");
	sfh_argv_add(&pathc, &pathv, LAM_SYSCONFDIR);

	fullpath = sfh_path_find(filename, pathv, R_OK);
	sfh_argv_free(pathv);
	return(fullpath);
}

/*
 *	node_ranges
 *
 *	Function:	- fill buffer with list of nodes
 *			- compact nodes to ranges
 *			- formatting buffer assumed large enough
 *	Accepts:	- formatting buffer
 *			- initial node
 *			- initial index
 */
static void
node_ranges(char *buf, int4 initnode, int4 initindex)
{
	int4		node;			/* target node */
	int4		index;			/* for getnode() */
	int4		flags;			/* for getnode() */
	int4		next;			/* next expected node */
	int		fl_flush = 0;		/* flush node flag */

	sprintf(smallbuf, "n%d", initnode);
	strcat(buf, smallbuf);
/*
 * Loop over the nodes.
 */
	next = initnode + 1;

	nid_get(&index, &node, &flags);

	while (index != initindex) {
		if (node != next) {
			if (next > initnode + 1) {
/*
 * We have a "hole" ending a range of length >= 2.
 */
				sprintf(smallbuf, "-%d,%d", next - 1, node);
				strcat(buf, smallbuf);
			}
			else {
/*
 * We have a "hole" ending a range of length 1.
 */
				sprintf(smallbuf, ",%d", node);
				strcat(buf, smallbuf);
			}
			initnode = node;
			next = node + 1;
			fl_flush = 0;
		}
/*
 * Otherwise, keep incrementing.
 */
		else {
			++next;
			fl_flush = 1;
		}

		nid_get(&index, &node, &flags);
	}
/*
 * If we were still in a range, flush it out.
 */
	if (fl_flush) {
		sprintf(smallbuf, "-%d", next - 1);
		strcat(buf, smallbuf);
	}

	strcat(buf, " ");
}

/*
 *	signal_handler
 *
 *	Function:	- dooms application started by lamexec
 *	Accepts:	- signal
 */
static void
signal_handler(int sig)
{
	static int	numsigs = 0;		/* # times signal caught */
	int		pid;

	if (sig != SIGINT && sig != SIGTERM) 
	  return;
/*
 * Second time around just exit.
 */
	if (numsigs++ > 0) {
		exit(1);
	}

	if ((pid = fork()) < 0) {
		return;
	}
	else if (pid == 0) {
		lam_reset_pid_cache();
		if (kinit(PRCMD)) {
			exit(errno);
		}

		app_doom(job_n, job, -15);	/* deliver SIGTERM */
		kexit(0);
	}
}

/*
 *	error_cleanup
 *
 *	Function:	- try to clean up wait messages
 *			- this is not foolproof but better than nothing
 */
static void
error_cleanup(void)
{
    struct nmsg		msg;
/*
 * Wait a little while.
 */
    sleep(1);

    LAM_ZERO_ME(msg);
    msg.nh_event = (-lam_getpid()) & 0xBFFFFFFF;
    msg.nh_length = 0;
    msg.nh_flags = 0;
/*
 * Loop trying to receive wait messages.
 */
    while (1) {
	msg.nh_type = 1;
	if (ntry_recv(&msg))
	    break;
    }
}

static void
first_signal_handler(int sig) {
  static int numsig = 0;
  static char error_message[] = "\
******************** WARNING ************************\n\
This is a vulnerable region. Exiting the application\n\
now may lead to improper cleanup of temporary objects\n\
To exit the application, press Ctrl-C again\n\
******************** WARNING *************************\n";

  if (sig != SIGINT && sig != SIGTERM) {
     return;
   }
 /* He really wants to quit, call the generic signal handler and
  * Note: signal_handler() does not return
  */
   if (numsig++ > 0) {
     signal_handler (sig);
   }
  /* This is the first time we have caught the signal */
  else {
    write (1, error_message, sizeof (error_message));
    return;
  }
}

